### ISSUE TITLE
Quick question: what namespace should we use?

### ISSUE THREAD

**Issue by chentong319** [onnx/onnx-mlir#1174]
1. If onnx-mlir is treated as a project under mlir, we should put all the onnx-mlir code under ::mlir::onnxmlir name space and the onnx dialect under ::mlir::onnxmlir::onnx namespace. Or we can just use ::mlir::onnx and ::mlir::onnx::onnx like torch-mlid project.
2. If we just treat onnx dialect as a dialect under mlir, then onnx should be under ::mlir::onnx namespace.

I prefer option 1. Option 2 is also OK and requires less change.

**Comment by AlexandreEichenberger** [comment]
I like your suggestion. I am not sure I can appreciate the difference between having a flat "mlir" and "onnxmlir" vs "mlir" and "mlir::onnxmlir". I would lean towards what other project did on this one.

**Comment by etiotto** [comment]
> If onnx-mlir is treated as a project under mlir, we should put all the onnx-mlir code under ::mlir::onnxmlir name space and the onnx dialect under ::mlir::onnxmlir::onnx namespace. Or we can just use ::mlir::onnx and ::mlir::onnx::onnx like torch-mlid project.

`onnx_mlir` **uses** MLIR but is not a project **under** `mlir` because it is not in tree. For that reason, I think we should not place our namespace `onnx_mlir` under `mlir` at this point.

My preference would be to use `onnx_mlir` as the top level namespace  and nest the `krnl`, `zhigh` and `zlow` dialects in that namespace (already done).  Calling the top level namespace `onnxmlir` rather than `onnx_mlir` is fine, however it would cause some "churn" and I am not sure is worth it. 

I agree with you that it would be nice to have a namespace fo the ONNX dialect. I would be fine with `onnx_mlir::onnx` for that dialect and related conversion passes. 

In summary:
  - onnx_mlir to be the top level namespace in the project. 
  - onnx_mlir::krnl, onnx_mlir::zhigh, onnx_mlir::zlow and onnx_mlir::onnx  to be the namespaces for, respectively, the `Krnl`, `ZHigh`, `ZLow`, `ONNX` dialect and related conversion passes  (mostly done, except for the ONNX operations)

Utilities specific to a dialect or dialect conversion pass should IMO be placed in that dialect namespace. 

Common utilities that are used by more than one dialect could be placed:
  - directly under the `onnx_mlir` namespace, or  (my slight personal preference)
  - under `onnx_mlir::utilities` (or something similar)

**Comment by etiotto** [comment]
As for the use of namespaces, I have been trying to follow these recommendations:
  - [never add a using declaration in a header file](https://stackoverflow.com/questions/5849457/using-namespace-in-c-headers)
  - https://stackoverflow.com/questions/8681714/correct-way-to-define-c-namespace-methods-in-cpp-file
  - https://stackoverflow.com/questions/10816600/c-namespaces-how-to-use-in-header-and-source-files-correctly

**Comment by AlexandreEichenberger** [comment]
Ok, so the suggestion is that 

1. `onnx_mlir` is a top level namespace (not under `mlir`). That is sensible.
2. each dialect has a sub-namespace under `onnx_mir`
3. we may use an `onnx_mlir::utilities` for utilities used under more than one dialect. I see that only as useful if we were to reuse names between top level and utilities... so it may not much sense.

What about accelerators? Should we have `onnx_mlir::accel::zhigh` to encapsulate all accelerator code under a common namespace? Or `accel_zhigh` ? That might be nice if we get more accelerators in the long term.

I fully agree that we should have no `using namespace` in a header file, we had some and I believe we removed them. I don't see much difference with a scoped namespace in a cpp file when in cover the entire file, so I believe that `using namespace onnx_mlir` should be used if the scope is the entire file.

Can you comments on what is still being worked on and what is currently complete? Thanks.

**Comment by chentong319** [comment]
Agree on using onnx_mlir as a top name space and  onnx_mlir::dialectname for dialects.  Will read the document for the coding details, such as when should we rely on using statement or use the namespace directly on objects.

**Comment by AlexandreEichenberger** [comment]
Is this complete, or do we still have work to do to update the current code base?

**Comment by etiotto** [comment]
>> What about accelerators? Should we have onnx_mlir::accel::zhigh 
We have been using `onnx_mlir::zhigh` for that dialect rather than `onnx_mlir::accel::zhigh`. The former ensures that the ZHigh dialect declarations belong to its own namespace under `onnx_mlir`. The later is accomplishes the same goal, but if we want to go that way we should  nest zhigh inside the particular accelerator namespace, so `onnx_mlir::accel::nnpa::zhigh`. At this point in time I think that is perhaps overkill.

**Comment by etiotto** [comment]
> Is this complete, or do we still have work to do to update the current code base?

We have made quite a bit of progress (KrnlToAffin and KrnlToLLVM have their namespaces), but this is not yet complete. Most noticeably the ONNX, and its shape inference code does not yet follows the namespace conventions we discussed above.
